src/charm.py

```diff
@@ -131,6 +133,37 @@ def __init__(self, *args):
             getattr(self.on, "cluster_relation_departed"), self._on_cluster_relation_changed
         )
 
+    @property
+    def _layer(self) -> Layer:
```

Comment:
**suggestion:** I'm fine with going with whatever you prefer on this, I'll just present a counter-point.

IMO, although this is K8s specific, it's just 'config' in one form or another. Having it in 'charm' might get in the way of understanding the actual even handling.
That said, it's not too important and it's absolutely fine to keep here if you prefer it, at least it's the same as Kafka.

Reply:
All layer interactions are done on `charm.py`, as that's where we have the main logic around start|restart (`init_server`, `pebble_layer_ready`, `_restart`)

Reply:
This really looks something that is `workload` specific. 

If I can provide a suggestion, I would probably move this to `workload.py`, this way you don't need to pass the layer to the `start` method. I would not pass the layer to `start` also because the signature of the method in the abstract class (defining the interface) does not have that argument, so we are breaking a bit the API. Beside, Layer objects are something really k8s specific, while workload should be generic as they apply to VM as well.

Looking at the definition, it looks that the only external informations to define the layer are  `self.state.unit_server.host` and `self.config_manager.server_jvmflags`. 

This is what I would probably do:

* for `self.state.unit_server.host`: provide this at the constructor of the workload. 
* for `SERVER_JVMFLAGS`: we could have two options: either (1) don't provide the environment in the layer, but rather writing the `/etc/environ` as we do for VM OR (2) add a envs variable to the start method (that could also be implemented for VM workloads)